From f89ee705eb8ce0f1be80effbd349ad5bcda40851 Mon Sep 17 00:00:00 2001
From: Bin Han <bin.b.han@intel.com>
Date: Thu, 18 Feb 2016 16:23:35 +0800
Subject: [PATCH 2/7] atomisp2: do not hold mutex when flushing work thread

In low CPU freq, the flushing work thread in atomisp_streamon
can take more than 1 second which blocks atomisp_isr_thread.

The flush actually does not need take the mutex as the work thread
itself does not really take the mutex to protect itself when it is
running. So release the mutex before flushing.

Change-Id: I790c9730794aab582a32bd8bb84942daf62444bd
Original-Tracked_On: #H1504159428
Signed-off-by: Bin Han <bin.b.han@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/102382
Reviewed-by: HaiqiangX Yao <haiqiangx.yao@intel.com>
Approver: Jian Xu Zheng <jian.xu.zheng@intel.com>
Reviewed-by: WeiX Meng <weix.meng@intel.com>
Integrator: WeiX Meng <weix.meng@intel.com>
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-21584
Signed-off-by: Tao Wei <wei.tao@intel.com>
Reviewed-on: https://android.intel.com:443/475694
---
 .../camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c    | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c
index f0a4edb..e36be1d 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c
@@ -1816,8 +1816,8 @@ static int atomisp_streamon(struct file *file, void *fh,
 					asd->latest_preview_exp_id);
 
 			if (asd->delayed_init == ATOMISP_DELAYED_INIT_QUEUED) {
-				flush_work(&asd->delayed_init_work);
 				rt_mutex_unlock(&isp->mutex);
+				flush_work(&asd->delayed_init_work);
 				if (wait_for_completion_interruptible(
 						&asd->init_done) != 0)
 					return -ERESTARTSYS;
-- 
1.9.1

